### PR TITLE
ElectLeaders api(KIP-460) implemented

### PR DESCRIPTION
**Pull request by PratRanj07**
Implemented Elect Leaders API and added unit tests for it.

### CODE REVIEW

File: examples/adminapi.py [hunk 1]
```diff
@@ -917,6 +944,7 @@ def example_delete_records(a, args):
         sys.stderr.write(' list_offsets <isolation_level> <topic1> <partition1> <offset_spec1> ' +
                          '[<topic2> <partition2> <offset_spec2> ..]\n')
         sys.stderr.write(' delete_records <topic1> <partition1> <offset1> [<topic2> <partition2> <offset2> ..]\n')
+        sys.stderr.write(' elect_leaders <election_type> <topic1> <partition1> <topic2> <partition2>...\n')
```
Comment (pranavrth) [inline examples/adminapi.py:947]:
```suggestion
        sys.stderr.write(' elect_leaders <election_type> <topic1> <partition1> [<topic2> <partition2> ..]\n')
```

File: src/confluent_kafka/admin/__init__.py [hunk 2]
```diff
@@ -1258,3 +1274,39 @@ def delete_records(self, topic_partition_offsets, **kwargs):
...
+        :rtype: rtype: dict[TopicPartition, future]
+
+        :raises KafkaException: Operation failed locally or on broker.
+        :raises TypeError: Invalid input type.
+        :raises ValueError: Invalid input value.
+        """
+
+        AdminClient._check_elect_leaders(election_type, partitions)
+
+        f, futmap = AdminClient._make_futures_v2(partitions, _TopicPartition, AdminClient._make_futmap_result)
```
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1308]: Input topic partitions can be None as well. In this case, we have to make future in different way. Please check https://github.com/confluentinc/confluent-kafka-python/blob/master/src/confluent_kafka/admin/__init__.py#L1148-L1152 for reference on how to make that work.

**Review by pranavrth — COMMENTED** [review]
One of the major comment. Still reviewing other files.

File: src/confluent_kafka/src/Admin.c [hunk 3]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
         "\n"
         "  This method should not be used directly, use confluent_kafka.AdminClient.delete_records()\n");
 
+/**
+ * @brief Elect leaders
+ */
+PyObject *Admin_elect_leaders(Handle *self, PyObject *args, PyObject *kwargs) {
+        PyObject *election_type = NULL, *partitions = NULL, *future;
+        rd_kafka_ElectLeaders_t *elect_leaders = NULL;
+        rd_kafka_ElectionType_t elec;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3068]: `c_election_type`

File: src/confluent_kafka/src/Admin.c [hunk 4]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+        }
+
+        /* options_to_c() sets future as the opaque, which is used in the
+         * background_event_cb to set the results on the future as the
+         * admin operation is finished, so we need to keep our own refcount. */
+        Py_INCREF(future);
+
+        elec = (rd_kafka_ElectionType_t)cfl_PyInt_AsInt(election_type);
+
+        if (!PyList_Check(partitions)) {
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3101]: Check None case as well.

File: src/confluent_kafka/src/Admin.c [hunk 5]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
         "\n"
         "  This method should not be used directly, use confluent_kafka.AdminClient.delete_records()\n");
 
+/**
+ * @brief Elect leaders
+ */
+PyObject *Admin_elect_leaders(Handle *self, PyObject *args, PyObject *kwargs) {
+        PyObject *election_type = NULL, *partitions = NULL, *future;
+        rd_kafka_ElectLeaders_t *elect_leaders = NULL;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3067]: `c_elect_leaders`

File: src/confluent_kafka/src/Admin.c [hunk 6]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+        if (!PyList_Check(partitions)) {
+                PyErr_SetString(PyExc_ValueError, "partitions must be a list");
+                goto err;
+        }
+        if (partitions != Py_None) {
+                c_partitions = py_to_c_parts(partitions);
+        }
+
+        elect_leaders = rd_kafka_ElectLeaders_new(elec, c_partitions);
+        rd_kafka_topic_partition_list_destroy(c_partitions);
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3110]: line break

File: src/confluent_kafka/src/Admin.c [hunk 7]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+        rd_kafka_ElectLeaders(self->rk, elect_leaders, c_options, rkqu);
+        CallState_end(self, &cs);
+
+        rd_kafka_queue_destroy(rkqu); /* drop reference from get_background */
+
+        rd_kafka_AdminOptions_destroy(c_options);
+        rd_kafka_ElectLeaders_destroy(elect_leaders);
+
+        Py_RETURN_NONE;
+err:
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3134]: line break

File: src/confluent_kafka/src/Admin.c [hunk 8]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+
+        if (c_options) {
+                rd_kafka_AdminOptions_destroy(c_options);
+                Py_DECREF(future);
+        }
+        return NULL;
+}
+
+const char Admin_elect_leaders_doc[] = PyDoc_STR(
+    ".. py:function:: elect_leaders(election_type, topic_partition_list, "
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3147]:
```suggestion
    ".. py:function:: elect_leaders(election_type, partitions, "
```

File: src/confluent_kafka/src/Admin.c [hunk 9]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+    "future, [request_timeout, operation_timeout])\n"
+    "\n"
+    "  Perform Preffered or Unclean election for the specified "
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3150]:
```suggestion
    "  Perform Preferred or Unclean election for the specified "
```

File: src/confluent_kafka/src/Admin.c [hunk 10]
```diff
@@ -3059,6 +3059,100 @@ const char Admin_delete_records_doc[] = PyDoc_STR(
...
+    "Topic+Partitions.\n"
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:3151]:
```suggestion
    "Topic Partitions.\n"
```

File: src/confluent_kafka/src/Admin.c [hunk 11]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
         return NULL;
 }
 
+static PyObject *Admin_c_ElectLeadersResult_to_py(
+    const rd_kafka_topic_partition_result_t **partitions,
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4625]: `partition_results` better.

File: src/confluent_kafka/src/Admin.c [hunk 12]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
...
+                const rd_kafka_topic_partition_t *rktpar;
+                const rd_kafka_error_t *error;
+
+                rktpar =
+                    rd_kafka_topic_partition_result_partition(partitions[i]);
+                error = rd_kafka_topic_partition_result_error(partitions[i]);
+                value = KafkaException_new_or_none(rd_kafka_error_code(error),
+                                                   rd_kafka_error_string(error));
+               
+                PyDict_SetItem(result, c_part_to_py(rktpar), value);
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4643]: Create new variable for `c_part_to_py(rktpar)`. Decrease reference for the created variable. Check other places.

File: src/confluent_kafka/src/Admin.c [hunk 13]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
...
+    const rd_kafka_topic_partition_result_t **partitions,
+    size_t cnt) {
+        PyObject *result = NULL;
+        size_t i;
+
+        result = PyDict_New();
+
+        for (i = 0; i < cnt; i++) {
+                PyObject *value = NULL;
+                const rd_kafka_topic_partition_t *rktpar;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4634]: `c_topic_partition`

File: src/confluent_kafka/src/Admin.c [hunk 14]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
...
+                const rd_kafka_error_t *error;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4635]: `c_error`

File: src/confluent_kafka/src/Admin.c [hunk 15]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
...
+
+                rktpar =
+                    rd_kafka_topic_partition_result_partition(partitions[i]);
+                error = rd_kafka_topic_partition_result_error(partitions[i]);
+                value = KafkaException_new_or_none(rd_kafka_error_code(error),
+                                                   rd_kafka_error_string(error));
+               
+                PyDict_SetItem(result, c_part_to_py(rktpar), value);
+                Py_XDECREF(value);
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4644]:
```suggestion
                Py_DECREF(value);
```

File: src/confluent_kafka/src/Admin.c [hunk 16]
```diff
@@ -4523,6 +4621,32 @@ static PyObject *Admin_c_DeletedRecords_to_py (const rd_kafka_topic_partition_li
...
+static PyObject *Admin_c_ElectLeadersResult_to_py(
+    const rd_kafka_topic_partition_result_t **partitions,
+    size_t cnt) {
+        PyObject *result = NULL;
+        size_t i;
+
+        result = PyDict_New();
+
+        for (i = 0; i < cnt; i++) {
+                PyObject *value = NULL;
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:4633]: No need to initialize

File: src/confluent_kafka/src/Admin.c [hunk 17]
```diff
@@ -4875,6 +4999,24 @@ static void Admin_background_event_cb (rd_kafka_t *rk, rd_kafka_event_t *rkev,
...
+                const rd_kafka_ElectLeaders_result_t
+                    *c_elect_leaders_res_event =
+                        rd_kafka_event_ElectLeaders_result(rkev);
+                const rd_kafka_ElectLeadersResult_t *c_elect_leaders_res =
+                    rd_kafka_ElectLeaders_result(c_elect_leaders_res_event);
+
+                const rd_kafka_topic_partition_result_t *
+                    *c_election_result_partitions =
+                        rd_kafka_ElectLeadersResult_partitions(
+                            c_elect_leaders_res, &c_result_cnt);
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:5013]: Do this inside `Admin_c_electLeadersResult_to_py` function.

File: src/confluent_kafka/src/Admin.c [hunk 18]
```diff
@@ -4875,6 +4999,24 @@ static void Admin_background_event_cb (rd_kafka_t *rk, rd_kafka_event_t *rkev,
...
+
+                result = Admin_c_ElectLeadersResult_to_py(
```
Comment (pranavrth) [inline src/confluent_kafka/src/Admin.c:5015]: Take `c_elect_leaders_res` as input.

File: src/confluent_kafka/src/confluent_kafka.h [hunk 19]
```diff
@@ -185,6 +185,7 @@ cfl_PyUnistr_AsUTF8 (PyObject *o, PyObject **uobjp) {
  ****************************************************************************/
 extern PyObject *KafkaException;
 
+PyObject *KafkaException_new_or_none (rd_kafka_resp_err_t err, const char *str);
```
Comment (pranavrth) [inline src/confluent_kafka/src/confluent_kafka.h:188]: Move this to KafkaException related functions.

File: src/confluent_kafka/src/confluent_kafka.c [hunk 20]
```diff
@@ -340,6 +340,32 @@ static PyTypeObject KafkaErrorType = {
...
 
+/**
+ * @brief Creates a KafkaException from error code and error string.
+ */
+PyObject *KafkaException_new_or_none (rd_kafka_resp_err_t err, const char *str) {
+        if (err) {
+
+                PyObject *excargs = NULL, *exc = NULL;;
+                PyObject *error = KafkaError_new0(err, str);
+                PyObject *exctype = KafkaException;
```
Comment (pranavrth) [inline src/confluent_kafka/src/confluent_kafka.c:351]: You should use KafkaException directly instead.

File: src/confluent_kafka/src/confluent_kafka.c [hunk 21]
```diff
@@ -340,6 +340,32 @@ static PyTypeObject KafkaErrorType = {
...
+                excargs = PyTuple_New(1);
+                PyTuple_SetItem(excargs, 0, error);
+                exc = ((PyTypeObject *)exctype)->tp_new(
+                (PyTypeObject *)exctype, NULL, NULL);
+
+                exc->ob_type->tp_init(exc, excargs, NULL);
+
+
+                Py_DECREF(excargs);
+                Py_XDECREF(error);
```
Comment (pranavrth) [inline src/confluent_kafka/src/confluent_kafka.c:361]: Py_DECREF

File: src/confluent_kafka/src/confluent_kafka.c [hunk 22]
```diff
@@ -340,6 +340,32 @@ static PyTypeObject KafkaErrorType = {
         KafkaError_new             /* tp_new */
 };
 
+/**
+ * @brief Creates a KafkaException from error code and error string.
+ */
+PyObject *KafkaException_new_or_none (rd_kafka_resp_err_t err, const char *str) {
+        if (err) {
+
+                PyObject *excargs = NULL, *exc = NULL;;
```
Comment (pranavrth) [inline src/confluent_kafka/src/confluent_kafka.c:349]: No need to initialize

File: src/confluent_kafka/src/confluent_kafka.c [hunk 23]
```diff
@@ -340,6 +340,32 @@ static PyTypeObject KafkaErrorType = {
...
+                PyObject *error = KafkaError_new0(err, str);
+                PyObject *exctype = KafkaException;
+                excargs = PyTuple_New(1);
+                PyTuple_SetItem(excargs, 0, error);
+                exc = ((PyTypeObject *)exctype)->tp_new(
+                (PyTypeObject *)exctype, NULL, NULL);
+
+                exc->ob_type->tp_init(exc, excargs, NULL);
+
```
Comment (pranavrth) [inline src/confluent_kafka/src/confluent_kafka.c:358]: Remove extra line break.

File: tests/test_Admin.py [hunk 24]
```diff
@@ -1194,3 +1195,34 @@ def test_delete_records():
...
+    a = AdminClient({"socket.timeout.ms": 10})
+
+    correct_topic_partition = TopicPartition("test-topic1", 0)
+    incorrect_partition = TopicPartition("test-topic1", -1)
+
+    correct_election_type = ElectionType.PREFERRED
+
+    # Incorrect Election Type
+    with pytest.raises(TypeError):
+        a.elect_leaders(None, [correct_topic_partition])
```
Comment (pranavrth) [inline tests/test_Admin.py:1210]: Also check with some other value like string as first argument.

File: tests/test_Admin.py [hunk 25]
```diff
@@ -1194,3 +1195,34 @@ def test_delete_records():
...
+
+    # Incorrect Partitions type
+    with pytest.raises(TypeError, match="Expected partitions to be a list, got 'str'"):
+        a.elect_leaders(correct_election_type, "1")
+
+    # Request-specific tests
```
Comment (pranavrth) [inline tests/test_Admin.py:1216]: What does this mean?

**Review by pranavrth — CHANGES_REQUESTED** [review]

File: tests/test_Admin.py (hunk 25 above)
Comment (PratRanj07) [reply tests/test_Admin.py]: Changed it to Partition-specific tests.
It is specifically for partitions array validation checks

File: examples/adminapi.py [hunk 26]
```diff
@@ -878,6 +878,39 @@ def example_delete_records(a, args):
...
+        raise ValueError(f"Invalid election_type: {args[0]}, expected 'PREFERRED' or 'UNCLEAN'")
+
+    for topic, partition in zip(args[1::2], args[2::2]):
+        partitions.append(TopicPartition(topic, int(partition)))
+
+    f = a.elect_leaders(election_type, partitions)
+    try:
+        results = f.result()
+        for partition, exception in results.items():
+
```
Comment (pranavrth) [inline examples/adminapi.py:903]: Remove extra line break.

File: examples/adminapi.py [hunk 27]
```diff
@@ -878,6 +878,39 @@ def example_delete_records(a, args):
...
+            if exception is None:
+                print(f"Leader Election Successful for topic {partition.topic} partition {partition.partition}")
```
Comment (pranavrth) [inline examples/adminapi.py:905]:
```suggestion
                print(f"Leader Election Successful for topic: '{partition.topic}' partition: '{partition.partition}'")
```

File: examples/adminapi.py [hunk 28]
```diff
@@ -878,6 +878,39 @@ def example_delete_records(a, args):
...
+            else:
+                print(
+                    "Leader Election Failed for topic " +
+                    f"{partition.topic} partition {partition.partition}: {exception}")
```
Comment (pranavrth) [inline examples/adminapi.py:909]:
```suggestion
                    f"'{partition.topic}' partition '{partition.partition}': {exception}")
```

File: examples/adminapi.py [hunk 29]
```diff
@@ -878,6 +878,39 @@ def example_delete_records(a, args):
...
+    f = a.elect_leaders(election_type, partitions)
+    try:
+        results = f.result()
+        for partition, exception in results.items():
+
+            if exception is None:
+                print(f"Leader Election Successful for topic {partition.topic} partition {partition.partition}")
+            else:
+                print(
+                    "Leader Election Failed for topic " +
```
Comment (pranavrth) [inline examples/adminapi.py:908]:
```suggestion
                    "Leader Election Failed for topic: " +
```

File: examples/adminapi.py [hunk 30]
```diff
@@ -917,6 +950,7 @@ def example_delete_records(a, args):
         sys.stderr.write(' list_offsets <isolation_level> <topic1> <partition1> <offset_spec1> ' +
                          '[<topic2> <partition2> <offset_spec2> ..]\n')
         sys.stderr.write(' delete_records <topic1> <partition1> <offset1> [<topic2> <partition2> <offset2> ..]\n')
+        sys.stderr.write(' elect_leaders <election_type> <topic1> <partition1> [<topic2> <partition2> ..]\n')
```
Comment (pranavrth) [inline examples/adminapi.py:953]: Now all topic partition are optional. So change the usages accordingly.

```suggestion
        sys.stderr.write(' elect_leaders <election_type> [<topic1> <partition1> <topic2> <partition2> ..]\n')
```

File: src/confluent_kafka/admin/__init__.py [hunk 31]
```diff
@@ -548,6 +550,21 @@ def _check_delete_records(request):
             if req.partition < 0:
                 raise ValueError("'partition' cannot be negative")
 
+    @staticmethod
+    def _check_elect_leaders(election_type, partitions):
+        if not isinstance(election_type, ElectionType):
+            raise TypeError("Expected election_type to be of type 'ElectionType'")
```
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:556]:
```suggestion
            raise TypeError("Expected 'election_type' to be of type 'ElectionType'")
```

**Review by pranavrth — CHANGES_REQUESTED** [review]
Mainly nits.

File: src/confluent_kafka/admin/__init__.py [hunk 32]
```diff
@@ -1258,3 +1276,38 @@ def delete_records(self, topic_partition_offsets, **kwargs):
...
+
+    def elect_leaders(self, election_type, partitions=None, **kwargs):
+        """
+        Perform Preferred or Unclean leader election for
+        all the specified topic partitions.
+
+        :param election_type: ElectionType - The type of election to perform.
+        :param partitions: List[TopicPartition]|None - The topic partitions to perform
+               the election on. Use ``None`` to perform on all the topic partitions.
+        :param float request_timeout: The overall request timeout in seconds,
```
Comment (pranavrth) [inline src/confluent_kafka/admin/__init__.py:1288]:
```suggestion
        :param ElectionType election_type: The type of election to perform.
        :param List[TopicPartition]|None partitions: The topic partitions to perform
                                                            the election on. Use ``None`` to perform on
                                                            all the topic partitions.
        :param float request_timeout: The overall request timeout in seconds,
```

**Review by pranavrth — CHANGES_REQUESTED** [review]
Couple of more nits. Otherwise looks good.

**Review by pranavrth — APPROVED** [review]
Thanks Pratyush!. LGTM!

**Pull request merged**